providers/efa/verbs.c

```diff
@@ -0,0 +1,1116 @@
+
+	if (recv_cq == send_cq && recv_cq) {
+		pthread_spin_unlock(&recv_cq->lock);
+	} else {
+		if (recv_cq)
+			pthread_spin_unlock(&recv_cq->lock);
+		if (send_cq)
+			pthread_spin_unlock(&send_cq->lock);
+	}
+}
+
+static int efa_check_qp_attr(struct efa_dev *dev,
+			     struct ibv_qp_init_attr *attr)
+{
+	if (!attr->recv_cq || !attr->send_cq)
+		return EINVAL;
+
+	if (attr->srq)
+		return EINVAL;
+
+	return 0;
+}
+
+static int efa_check_qp_limits(struct efa_dev *dev,
+			       struct ibv_qp_init_attr *attr)
+{
+	if (attr->cap.max_send_sge > dev->max_sq_sge)
+		return EINVAL;
+
+	if (attr->cap.max_recv_sge > dev->max_rq_sge)
+		return EINVAL;
+
+	if (attr->cap.max_send_wr > dev->max_sq_wr)
+		return EINVAL;
+
+	if (attr->cap.max_recv_wr > dev->max_rq_wr)
+		return EINVAL;
+
+	return 0;
+}
+
+static struct ibv_qp *create_qp(struct ibv_pd *ibvpd,
+				struct ibv_qp_init_attr *attr,
+				uint32_t driver_qp_type)
+{
+	struct efa_context *ctx = to_efa_context(ibvpd->context);
+	struct efa_dev *dev = to_efa_dev(ibvpd->context->device);
+	struct efa_create_qp_resp resp = {};
+	struct efa_create_qp req = {};
+	struct efa_cq *send_cq;
+	struct efa_cq *recv_cq;
+	struct efa_qp *qp;
+	int err;
+
+	err = efa_check_qp_attr(dev, attr);
+	if (err)
+		return NULL;
+
+	err = efa_check_qp_limits(dev, attr);
+	if (err)
+		return NULL;
+
+	qp = calloc(1, sizeof(*qp));
+	if (!qp)
+		return NULL;
+
+	efa_setup_qp(qp, &attr->cap, dev->pg_sz);
+
+	attr->cap.max_send_wr = qp->sq.wq.wqe_cnt;
+	attr->cap.max_recv_wr = qp->rq.wq.wqe_cnt;
+
+	req.rq_ring_size = (qp->rq.wq.desc_mask + 1) *
+		sizeof(struct efa_io_rx_desc);
+	req.sq_ring_size = (attr->cap.max_send_wr) *
+		sizeof(struct efa_io_tx_wqe);
+	if (attr->qp_type == IBV_QPT_DRIVER)
+		req.driver_qp_type = driver_qp_type;
+
+	if (ibv_cmd_create_qp(ibvpd, &qp->ibvqp, attr, &req.ibv_cmd,
+			      sizeof(req), &resp.ibv_resp, sizeof(resp)))
+		goto err_free_qp;
+
+	qp->ibvqp.state = IBV_QPS_RESET;
+	qp->sq_sig_all = attr->sq_sig_all;
+	qp->ctx = ctx;
+
+	err = efa_rq_initialize(qp, &resp);
+	if (err)
+		goto err_destroy_qp;
+
+	err = efa_sq_initialize(qp, &resp);
+	if (err)
+		goto err_terminate_rq;
+
+	pthread_spin_lock(&ctx->qp_table_lock);
+	ctx->qp_table[qp->ibvqp.qp_num] = qp;
+	pthread_spin_unlock(&ctx->qp_table_lock);
+
+	if (attr->send_cq) {
+		send_cq = to_efa_cq(attr->send_cq);
+		qp->scq = send_cq;
+		pthread_spin_lock(&send_cq->lock);
+		efa_cq_inc_ref_cnt(send_cq, resp.send_sub_cq_idx);
+		pthread_spin_unlock(&send_cq->lock);
+	}
+
+	if (attr->recv_cq) {
+		recv_cq = to_efa_cq(attr->recv_cq);
+		qp->rcq = recv_cq;
+		pthread_spin_lock(&recv_cq->lock);
+		efa_cq_inc_ref_cnt(recv_cq, resp.recv_sub_cq_idx);
+		pthread_spin_unlock(&recv_cq->lock);
+	}
+
+	return &qp->ibvqp;
+
+err_terminate_rq:
+	efa_rq_terminate(qp);
+err_destroy_qp:
+	ibv_cmd_destroy_qp(&qp->ibvqp);
+err_free_qp:
+	free(qp);
+	return NULL;
+}
+
+struct ibv_qp *efa_create_qp(struct ibv_pd *ibvpd,
+			     struct ibv_qp_init_attr *attr)
```

Comment:
You have to confirm that the passed ibbvpd is an efa pd, not just crash.

Reply:
I don't understand, what else could it be?
Do you want the efa_context to remember all allocated efa_pd's?

Reply:
It could be any drivers' ibvpd, once you are in dv land there is no protection that the driver is being called appropriately.

The mlx5 driver does this check

Reply:
OK, I was confused as this is the standard verb function.
I will add a check to efadv_create_driver_qp.